ooniapi/services/ooniprobe/src/ooniprobe/routers/v1/probe_services.py

```diff
@@ -590,3 +565,234 @@ def random_web_test_helpers(th_list: List[str]) -> List[Dict]:
+    session: PostgresSessionDep,
+    content_encoding: str = Header(default=None),
+) -> SubmitMeasurementResponse | Dict[str, Any]:
+    """
+    Submit measurement
+    """
+    setnocacheresponse(response)
+    empty_measurement = {}
+    try:
+        rid_timestamp, test_name, cc, asn, format_cid, rand = report_id.split("_")
+    except Exception:
+        log.info("Unexpected report_id %r", report_id[:200])
+        raise error("Incorrect format")
+
+    # TODO validate the timestamp?
+    good = len(cc) == 2 and test_name.isalnum() and 1 < len(test_name) < 30
+    if not good:
+        log.info("Unexpected report_id %r", report_id[:200])
+        error("Incorrect format")
+
+    try:
+        asn_i = int(asn)
+    except ValueError:
+        log.info("ASN value not parsable %r", asn)
+        error("Incorrect format")
+
+    if asn_i == 0:
+        log.info("Discarding ASN == 0")
+        Metrics.MSMNT_DISCARD_ASN0.inc()
+        return empty_measurement
+
+    if cc.upper() == "ZZ":
+        log.info("Discarding CC == ZZ")
+        Metrics.MSMNT_DISCARD_CC_ZZ.inc()
+        return empty_measurement
+
+    # Retrieve manifest known by the client
+    manifest = OONIProbeManifest.get_by_version(session, submit_request.manifest_version)
+    if manifest is None:
+        error({
+            "detail" : f"No manifest with version '{submit_request.manifest_version}' was found",
+            "error" : "manifest_not_found"
+        })
+
+    # Run verification
+    assert manifest
+    assert "probe_cc" in submit_request.content and isinstance(submit_request.content['probe_cc'], str)
+    assert "probe_asn" in submit_request.content and isinstance(submit_request.content['probe_asn'], str)
+
+    state: OONIProbeServerState = manifest.server_state
+    protocol_state = state.to_protocol()
+
+    try:
+        submit_response = protocol_state.handle_submit_request(
+            submit_request.nym,
+            submit_request.zkp_request,
+            submit_request.content["probe_cc"],
+            submit_request.content["probe_asn"],
+            list(submit_request.age_range),
+            list(submit_request.msm_range),
+            )
+        is_verified = True
+    except (DeserializationFailed, ProtocolError, CredentialError) as e:
+        # proof failed
+        # TODO Q: should we add a "why not verified" field to the measurement?
+        log.error(f"ZKP Failed: {e}")
+        is_verified = False
+        submit_response = None
+
+    data = submit_request.model_dump()
+
+    # Add verification-related data.
+    data['is_verified'] = is_verified
+    data_buff = io.BytesIO()
+    stream = io.TextIOWrapper(data_buff, "utf-8")
+    ujson.dump(data, stream)
+    data_bin = data_buff.getvalue()
+
+    # Write the whole body of the measurement in a directory based on a 1-hour
+    # time window
+    now = datetime.now(timezone.utc)
+    h = sha512(data_bin).hexdigest()[:16]
+    ts = now.strftime("%Y%m%d%H%M%S.%f")
```

Comment:
Can we use ISO standard? Make sure it has in it `T` and `Z` to indicate timezone.